src/core/Akka.Cluster/ClusterHeartbeat.cs

```diff
@@ -364,68 +395,137 @@ public ClusterHeartbeatSenderState HeartbeatRsp(UniqueAddress from)
+                        if (isUnreachable && acc.Count >= MonitoredByNumberOfNodes)
+                        {
+                            return take(n, iter, acc); // skip the unreachable, since we have already picked `MonitoredByNumberOfNodes`
+                        }
+                        else if (isUnreachable)
+                        {
+                            return take(n, iter, acc.Add(next)); // include the unreachable, but don't count it
+                        }
+                        else
+                        {
+                            return take(n - 1, iter, acc.Add(next)); // include the reachable
+                        }
+                    }
+                };
+
+                var tuple = take(MonitoredByNumberOfNodes, NodeRing().From(sender).Skip(1).GetEnumerator(), ImmutableSortedSet<UniqueAddress>.Empty);
+                var remaining = tuple.Item1;
+                var slice1 = tuple.Item2;
+
+                IImmutableSet<UniqueAddress> slice = remaining == 0 
+                    ? slice1 
+                    : take(remaining, NodeRing().Until(sender).Where(c => !c.Equals(sender)).GetEnumerator(), slice1).Item2;
+
+                return slice.ToImmutableHashSet();
             }
-            return slice.ToImmutableHashSet();
         }
 
-        public HeartbeatNodeRing Copy(UniqueAddress selfAddress = null, IEnumerable<UniqueAddress> nodes = null,
-            int? monitoredByNumberOfNodes = null)
+        public HeartbeatNodeRing Copy(UniqueAddress selfAddress = null, ImmutableHashSet<UniqueAddress> nodes = null, ImmutableHashSet<UniqueAddress> unreachable = null, int? monitoredByNumberOfNodes = null)
         {
-            return new HeartbeatNodeRing(selfAddress ?? SelfAddress,
-                nodes ?? NodeRing,
+            return new HeartbeatNodeRing(
+                selfAddress ?? SelfAddress,
+                nodes ?? Nodes,
+                unreachable ?? Unreachable,
                 monitoredByNumberOfNodes.HasValue ? monitoredByNumberOfNodes.Value : MonitoredByNumberOfNodes);
         }
 
         #region Operators
 
         public static HeartbeatNodeRing operator +(HeartbeatNodeRing ring, UniqueAddress node)
         {
-            return ring.NodeRing.Contains(node) ? ring : ring.Copy(nodes: ring.NodeRing.Add(node));
+            return ring.Nodes.Contains(node) ? ring : ring.Copy(nodes: ring.Nodes.Add(node));
         }
 
         public static HeartbeatNodeRing operator -(HeartbeatNodeRing ring, UniqueAddress node)
         {
-            return ring.NodeRing.Contains(node) ? ring.Copy(nodes: ring.NodeRing.Remove(node)) : ring;
+            return ring.Nodes.Contains(node) || ring.Unreachable.Contains(node)
+                ? ring.Copy(nodes: ring.Nodes.Remove(node), unreachable: ring.Unreachable.Remove(node)) 
+                : ring;
         }
 
         #endregion
+
+        #region Comparer
+        internal class RingComparer : IComparer<UniqueAddress>
```

Comment:
Scala version has a bit different implementation, but is still compares only UIDs